Enable securitymanager

rmuir opened the pull request:
We have been using this in our tests for some time. Recently the permissions have gotten reasonable where we don't need read/write access everywhere.

I think we should start ES with securitymanager (just like tomcat and other things) to enhance security.

I ran the lucene benchmark suite comparing security manager off/on and there is no difference. I think a lot of concerns about perceived performance differences don't really make sense if you are managing files and sockets properly.

Here is a prototype, i tried to keep it simple (but damn if the SSD detection and other stuff we have going on fights me on that). There is nothing to configure, though security manager can be disabled in elasticsearch.yml or whatever with `security.enabled=false`. 

We use the same actual policy file for both tests and bin/elasticsearch. We just add additional permissions based on the environment (e.g. data paths and so on) so things will work. 

I didnt write any tests yet, but I can add some unit tests for the stuff here.

rmuir committed:
enable securitymanager

jpountz commented on src/main/java/org/elasticsearch/bootstrap/Bootstrap.java:
```diff
@@ -92,6 +91,16 @@ public boolean handle(int code) {
+            Security.configure(environment);
+            logger.info("security enabled");
+        } else {
+            logger.info("security disabled");
```
Should it be a warning?

rmuir replied:
If its disabled lets issue a warning.

rmuir commented on src/main/java/org/elasticsearch/bootstrap/Security.java:
```diff
@@ -0,0 +1,160 @@
+                addRecursivePath(writer, "/sys/devices", "read,readlink");
+                addRecursivePath(writer, "/dev", "read,readlink");
+                addRecursivePath(writer, "/devices", "read,readlink");
+            }
```
I want to nuke this entire terrifying 'if' statement and just have the filestore info in Environment (with the other paths) instead of NodeEnvironment. I will open a separate PR for this.

rmuir committed:
Merge branch 'master' into put_me_in_coach

rmuir committed:
Remove crazy permissions for filestores, ssds, now that
this logic has been refactored.

Log a warning when security is disabled.

rmuir committed:
Merge branch 'master' into put_me_in_coach

Conflicts:
	src/main/java/org/elasticsearch/env/NodeEnvironment.java

rmuir committed:
ensure we only pull system filestores once time

rmuir commented:
I pushed commits simplifying the SSD/filestore stuff after #10755 and adding the warning as @jpountz suggested.

rmuir committed:
Ensure paths exist (or more permissions are needed later)

rmuir committed:
remove stray sop

s1monw commented:
this change looks good to me - i am curious if we see per drops with this enabled by default etc. but I even if we do we should push this (plus unittests please) and move the decision to default enable it to a different issue if we need to. I really like that we can offer this to the users and I am really leaning towards enable by default!

kimchy commented on src/main/java/org/elasticsearch/bootstrap/Bootstrap.java:
```diff
@@ -92,6 +91,16 @@ public boolean handle(int code) {
+            logger.info("security enabled");
+        } else {
+            logger.warn("security disabled");
+        }
```
can we debug log the default? also leaning to have info the "non default" setting, thats what we try to do most times in other components to try and keep the startup logs clean and informative.

rmuir replied:
This is officially back and forth at this point. I started at info and went to warn based on feedback. Guys get your shit together, im not playing this game.

mikemccand commented:
I tested indexing tiny (log line) docs, same test I run at http://benchmarks.elastic.co.

I ran each of master and this branch two times...

Master is 8999.7 and 9160.0 docs/sec overall (6.9 M docs), while this branch is 9610.9 and 9835.4 docs/sec.

Net/net I don't think security manager hurts indexing performance!

rmuir commented:
Unless someone can find real performance issues or problems, i plan to push the change in 72 hours. And you guys can bikeshed logging levels somewhere else, im not doing that.

kimchy commented:
Very promising on the perf test, its great. I will fix the logging post your push, I think that we should also work on a different issue on getting the test permissions out of this file, since it will be confusing for users who look at it, even though those env vars are not set.

rmuir commented:
i can remove the logging completely. i am happy with no logging and for someone else to deal with logging completely. its such a ridiculous thing to make bikesheds out of. I will not add any more logging statements to this codebase, trust me.

rmuir committed:
remove logging statements for another bikeshed

rmuir commented:
logging is removed.

kimchy commented:
LGTM, lets push it

rmuir committed:
Remove policy config file, its a resource.
Remove exposed boolean to turn off security.
Add unit test

rmuir commented:
I made the policy file a resource (not in conf/) and removed the docs of the option in elasticsearch.yml. Someone can turn it off with security.manager.enabled=false if they dont like it. And they can set JVM_OPTS with -D's to do whatever their heart desires rather than us supporting plumbing. I also added some really simple unit tests so we had something.

rmuir committed:
Buffer this inputstream out of paranoia. guava copy goes byte-by-byte...

s1monw commented on src/main/java/org/elasticsearch/bootstrap/Bootstrap.java:
```diff
@@ -92,6 +91,12 @@ public boolean handle(int code) {
     }
+    
+    private void setupSecurity(Settings settings, Environment environment) throws Exception {
+        if (settings.getAsBoolean("security.manager.enabled", true)) {
```
can we make `"security.manager.enabled"` a constant please

rmuir replied:
It is a constant. Just like all the other strings in this file :)

So I will add it, but I will do so under protest:
* there are zero named constants in this file, I am just matching the style surrounding the code, for consistency.
* named constants only used once just cause indirection hurt locality of reference to those reading the code.

In general, named constants considered harmful.
*

s1monw commented:
left one tiny nitpick - LGTM otherwise feel free to push after fixing the nitpick :)

kimchy commented:
LGTM, @rmuir if someone provides the `-Djava.security.policy` option, we will still run our code as far as I can tell, just double checking it won't be a problem?

rmuir commented:
Someone with a custom policy will have to also use the option (security.manager.enabled=false) to disable what we are doing.

kimchy commented:
@rmuir I see, will it be trappy? which one will be applied if just `-Djava.security.policy` is provided (not sure about the logic of which ones wins), maybe we should not do our thing if the system property exists as well?

rmuir commented:
I don't think we should have lots of logic to look if security manager is already set, to look if this or that system property is already set. There is no need or use-case for this IMO. It will just make our codebase even bigger than i already is.

Having a custom security policy is extremely expert. Those users can turn it off with the boolean i provided here. 

We only need one way to do this. This isn't perl.

kimchy commented:
@rmuir sounds good

rmuir committed:
Merge branch 'master' into put_me_in_coach

rmuir committed:
add constant only used once to make it harder to read the code

tlrx commented:
@rmuir sorry to come a day after the fair but I think this change prevents writing in a local FsRepository no?

rmuir commented:
Probably. I have no idea about this one in particular, but i am sure this change breaks all kinds of things using rogue paths etc. We should likely review all places using paths.get and open bugs for anything except environment that uses it.

tlrx commented:
I agree, it will surely breaks things here and there. But in the case of snapshot repositories of type `fs`, users provide a custom path where the snapshots will be written/read (see the [doc example](http://www.elastic.co/guide/en/elasticsearch/reference/current/modules-snapshots.html#_repositories))... I suppose it won't work unless users provide a custom policy file (and restart the JVM) or disable the security manager.

rmuir commented:
Why doesnt that feature use a path configured in environment like all other paths. That is the way to fix it imo

s1monw commented:
> Why doesnt that feature use a path configured in environment like all other paths. That is the way to fix it imo

I agree this should not be outside of the path or we need to have a dedicated path that is configured.

tlrx commented:
@rmuir @s1monw I agree and opened #10828 for that.
